### PR TITLE
wip: IR-402: Expose the publishing status of the cluster

### PR DESCRIPTION
**Pull request by flavianmissi**
Provide an API for cluster operators to consume the publishing status of the cluster at install time (Internal/External).

### CODE REVIEW

File: enhancements/installer/ipi-publishing-status.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
+
+* As an operator author, I want to easily discover the cluster publishing
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:36]: can we illustrate with a concrete example here?

Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: Thanks for reviewing @enxebre - I gave my own use case as example, let me know if it works.

File: enhancements/installer/ipi-publishing-status.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+* `Public`
+* `PublicAndPrivate`
+* `Private`
+
+`PublicAndPrivate` does not translate well to OpenShift. It allows public API
+server access and private node communication with the control plane, making
+the cluster appear public to end-users, while securing control plane and
+nodes communication.
+
+The suggested approach to Hypershift is to translate the modes as follows:
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:123]: what's the rationale for this suggestion? `Public`, `PublicAndPrivate` and `Private` are current options in hcp for aws, they represent cluster endpoints reachability for serving components like kas. Then ingress scope for the data plane is decoupled from the endpoints choice. Also fwiw most operator will run next to the kas being able to access it through the service network, unless they need to run operands in the data plane that need to talk back to the kas.
It'd be a good exercise to pick a few examples and see how they would look like in different form factors

Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:123]: TBH this is the most fragile part of this EP, because it's where I have the least understanding (hcp).
From my understanding of previous discussions with @csrwng, configuring an internal HCP cluster does not work the same way it does on OCP.
With HCP, there are three options, while with OCP there are two.
The way I see it these are our options: 
1. translate the choices so that Hypershift can set the Internal/External value on the Infrastructure object based on Public/PublicAndPrivate/Private;
2. change nothing and document the impact;

Option 2 works for me as it simplifies my work in this EP and my use case is currently specific to Azure. 
As long as the impact of exposing cluster publishing status is not restricted by our choice I'm fine with it. Notably, the openshift-apiserver as well as the ingress operator are currently configurable as internal/external at cluster bootstrap time depending on the user choice to publish the cluster or not.

TL;DR: the ultimate idea of this EP is to decentralize public vs private component configuration. If this is not something we want for Hypershift then I would just want to document it here.

File: enhancements/installer/ipi-publishing-status.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [IR-302](https://issues.redhat.com/browse/IR-302)
+---
+
+# Expose Cluster Publishing Status
+
+## Summary
+
+IPI clusters can be configured to be public or private. This is done in the
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:22]: can we define public and private? e.g does this refer to kas access, ingress for data plane, both, others?

File: enhancements/installer/ipi-publishing-status.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
+
+* As the author of the image registry operator, I want to easily discover the
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:36]: I'd expect cloud access to be granted only to the registry and I'd expect access to the registry to be dictated by the ingress controller setup, right? How does this new API knob intersect in that flow?

File: enhancements/installer/ipi-publishing-status.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+## Proposal
+
+The Infrastructure API is extended, adding the `Publish` field to the
+`InfrastructureStatus`.
+
+```go
+type InfrastructureStatus struct {
+	// ...
+
+	// publish indicates whether the cluster was installed as public or private.
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:60]: Why are we using two different terms public/external and private/internal to refer to the same thing? can we consistently use only one?

File: enhancements/installer/ipi-publishing-status.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	// The default is 'External', which means the cluster will be publicly
+    // available on the internet.
+	// The 'Internal' mode indicates a private cluster, and operators should
+    // configure operands accordingly.
+    // +kubebuilder:default=External
+    // +kubebuilder:validation:Enum=Internal;External
```
Comment (enxebre) [inline enhancements/installer/ipi-publishing-status.md:66]: ptal @JoelSpeed

File: enhancements/installer/ipi-publishing-status.md (hunk 3 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:22]: Gave it a shot. What I wrote is based on [OCP's docs](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/configuring-private-cluster.html).

File: enhancements/installer/ipi-publishing-status.md (hunk 5 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: Rephrased.

File: enhancements/installer/ipi-publishing-status.md (hunk 2 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:130]: Thanks for bringing this to light. I'm not aware of the specific use-cases for these different combinations, so I don't know how each of them should affect the registry - I'll need to discuss this with PM.
cc @dmesser

Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:130]: > Can we try to detail how this new field would impact the registry operator behaviour (following with your example) for these scenarios.

The registry operator would bootstrap the registry as private when the apiserver is also private. We consider its assets (backing storage) to be private. The registry endpoint would follow whatever ingress configuration the cluster has.

> Can we try to detail how the value/absence of this new field would impact the registry operator in the aws hypershift scenarios?

Currently the registry operator will not look at the cluster publishing status for AWS, so it will not be affected. If it would use this value, it should work as it would for the ARO example: when the apiserver is internal then so will be the assets the creates.

File: enhancements/installer/ipi-publishing-status.md (hunk 6 above)
Comment (JoelSpeed) [reply enhancements/installer/ipi-publishing-status.md:73]: To apply a default, you need to have `omitempty` and make the field optional with `// +optional`.

That said, what happens on upgrades here? If the field isn't already set you'll set it by default to External? That doesn't seem right. I think you need to add this with an empty default to support the upgrade case unless you can reliably correct this.

Is the field applicable to all platforms or only some platforms?

File: enhancements/installer/ipi-publishing-status.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+title: expose-cluster-publishing-status
+authors:
+  - "@flavianmissi"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@patrickdillon, installer"
+  - "@Miciah, edge networking"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@bparees"
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:11]: Should be me or David Eads, in this case you can assign me for it

File: enhancements/installer/ipi-publishing-status.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+### Goals
+
+Expose the cluster publishing status used at install time in the Infrastructure
+object.
+
+### Non-Goals
+
+This enhancement does not cover support for updating the cluster publishing
+status on running clusters. This property is meant to reflect the `publish`
+parameter as it was set in the install-config.yaml.
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:56]: Have you described somewhere how upgraded clusters will behave?

File: enhancements/installer/ipi-publishing-status.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+	// "External" indicates that the cluster is publicly available on the internet.
+	ExternalPublishMode PublishMode = "External"
+)
+```
+
+### Workflow Description
+
+1. The cluster creator install a cluster via IPI, setting `publish: Internal` in installer-config.yaml
+2. The installer bootstraps the cluster, and sets `Publish: Internal` in the Infrastructure status
+3. Operators use this value to configure operands accordingly.
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:99]: How many operators do you expect will actually need to be aware of this configuration?

File: enhancements/installer/ipi-publishing-status.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:107]: Boilerplate, can remove

File: enhancements/installer/ipi-publishing-status.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+The `Publish` field in the infrastructure status will only reflect the initial
+status of the cluster. Individual components (like ingress) may be changed on
+day-2 to for example turn a public cluster private, and the `Publish` field in
+the Infrastructure status will remain untouched.
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:112]: What's involved in converting a cluster from public to private? Is there a documented solution for this?

File: enhancements/installer/ipi-publishing-status.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+* `Public` -> `Public`
+* `PublicAndPrivate` -> `Public`
+* `Private` -> `Private`
+
+### Risks and Mitigations
+
+Users may interpret the `Publish` field in the Infrastructure status as a way
+to change the publishing status of their clusters at run time, which is not
+its intention.
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:140]: Users should not be updating the status, if you want to make the field immutable, that can be done with CEL or admission policy (the latter if you want to enable a particular user (eg a controller) to update the field, but no one else)

File: enhancements/installer/ipi-publishing-status.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+themselves as public or private according to the user request at install time.
+
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
```
Comment (JoelSpeed) [inline enhancements/installer/ipi-publishing-status.md:41]: I think it would be helpful to expand on the issues that the registry operator is having and how adding this new field will help to solve that, I don't really have a sense for that from reading this doc right now

File: enhancements/installer/ipi-publishing-status.md (hunk 11 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:112]: Day-2 conversion of a cluster from public to private involves manually configuring components individually. The process is described in [OCP docs[1]](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/configuring-private-cluster.html).

[1] https://docs.openshift.com/container-platform/4.13/post_installation_configuration/configuring-private-cluster.html

File: enhancements/installer/ipi-publishing-status.md (hunk 9 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:99]: I only expect registry operator to depend on this. We do have use cases for other parts of OCP that could move to this approach though. Namely ingress, DNS and apiserver. ATM I think they all rely on the installer to configure them as internal/external depending on what the user set in install-config.yaml.

File: enhancements/installer/ipi-publishing-status.md (hunk 12 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:140]: Can you give me a reference to CEL? first time I hear about it.
This field is not intended to be updated by anyone after it's written, so we don't need an admission policy at this point.

File: enhancements/installer/ipi-publishing-status.md (hunk 13 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:41]: gave it a shot

File: enhancements/installer/ipi-publishing-status.md (hunk 10 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: done

File: enhancements/installer/ipi-publishing-status.md (hunk 6 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:85]: You are correct, we need the default to be empty - gave that a shot but unsure about syntax, PTAL.

> Is the field applicable to all platforms or only some platforms?

Good question. While the `publish` option is widely available among supported platforms, I don't think it's available for _all_ of them. Perhaps the empty value is good enough for the ones where this does not apply?

File: enhancements/installer/ipi-publishing-status.md (hunk 7 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: Done

File: enhancements/installer/ipi-publishing-status.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  - "@patrickdillon, installer"
+  - "@Miciah, edge networking"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@JoelSpeed"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [IR-302](https://issues.redhat.com/browse/IR-302)
```
Comment (dhellmann) [inline enhancements/installer/ipi-publishing-status.md:15]:
```suggestion
  - https://issues.redhat.com/browse/IR-302
```

This part of the file is YAML embedded in the overall Markdown. That means you can't really use Markdown syntax directly. You can either just have the link, or if you want to try for Markdown you could wrap the value in quotes and see if that makes the YAML parser ignore it.

File: enhancements/installer/ipi-publishing-status.md (hunk 11 above)
Comment (JoelSpeed) [reply enhancements/installer/ipi-publishing-status.md:123]: It may be helpful to include that link within the EP.

Typically a publish field like this would be the source of truth for whether a cluster is public or private. Have the cluster half public and half private sounds like it could be confusing to understand and debug not only for end users, but also for support. Is this half and half scenario something we want to continue to support through this EP, or perhaps we need to breakdown the publishing status so there is still a source of truth, but it's per component.

Eg if the spec had an ingressController.Publish field that the ingress operator could use to determine if it should be public or private, that might make things very clear and automate some of these manual steps away

File: enhancements/installer/ipi-publishing-status.md (hunk 9 above)
Comment (JoelSpeed) [reply enhancements/installer/ipi-publishing-status.md:114]: Does the registry operator have its own specific configuration object? I wonder if placing this globally within the infrastructure, but not using it globally could be confusing for users

File: enhancements/installer/ipi-publishing-status.md (hunk 12 above)
Comment (JoelSpeed) [reply enhancements/installer/ipi-publishing-status.md:151]: Sure, take a look here, https://kubernetes.io/docs/reference/using-api/cel/

And as an example of an immutable field, https://github.com/openshift/api/blob/f404f2bc3524e03f3bc29a7b3cd0f624cd36eae7/example/v1/types_stable.go#L48-L58

File: enhancements/installer/ipi-publishing-status.md (hunk 6 above)
Comment (JoelSpeed) [reply enhancements/installer/ipi-publishing-status.md]: Typically where a field does not apply to all platforms, we put the field within the platformSpec or platformStatus discriminated unions. Of the IPI platforms we support, can you list which do and which do not support the publish option? That enumeration is probably good to include in the EP anyway btw so we know which platforms are affected by this change

Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: According to the [installer source code](https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L150-L156), changing the `publish` field in the installer-config.yaml to `Internal` is _supported_ on the following:
* AWS
* Azure
* GCP
* Alibaba
* IBM Cloud
* PowerVS

I think it would make sense to keep the `Publish` field platform specific as per your suggestion - changed the code, PTAL.

File: enhancements/installer/ipi-publishing-status.md (hunk 9 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:127]: Yes the registry operator has its own configuration object. 
We (me + installer folks) have discussed the possibility of having the installer configure the registry to internal when appropriate, but a decentralized approach made more sense at the time.
I do agree that it could be confusing to support both, the installer way (the way it currently works) and this decentralized way. I think that ideally we would request for the teams being configured by the installer to consume this value instead, so that eventually only the decentralized approach is used.

File: enhancements/installer/ipi-publishing-status.md (hunk 11 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md:136]: Yeah we might need to rename the field so that it's more clear that this is the initial publishing status (hey that might be a good name!) as to avoid confusion.

As for the half public and half private thing, it's more of a thing than I previously knew - ARO and Hypershift both support a combination of private and public along with the regular all public and all private installs. So this isn't only a day-2 thing. See https://issues.redhat.com/browse/CORS-2743.

File: enhancements/installer/ipi-publishing-status.md (hunk 12 above)
Comment (flavianmissi) [reply enhancements/installer/ipi-publishing-status.md]: thanks! applied to code sample.
